File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def mergeTwoLists(head1,head2):

    dummy = ListNode(0)

    temp = dummy

    while head1 and head2:

        if head1.val< head2.val:
            temp.next = ListNode(head1.val)
            temp = temp.next
            head1 = head1.next
        else:
            temp.next = ListNode(head2.val)
            temp = temp.next
            head2 = head2.next

    while head1:
        temp.next = ListNode(head1.val)
        temp = temp.next
        head1 = head1.next
    while head2:
        temp.next = ListNode(head2.val)
        temp = temp.next
        head2 = head2.next
    return dummy.next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
       
        if len(lists) == 0:
            return None
        if lists[0] == None:
            return None
        for i in range(1,len(lists)):
            lists[i] = mergeTwoLists(lists[i-1],lists[i])
        print(lists)
        return lists[len(lists)-1]
        

File: Data Structures & Algorithms/reorder-linked-list/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        if(head==null || head.next==null ) return ;
        ListNode slow=head;
        ListNode fast=head.next;

        while(fast.next!=null && fast.next.next!=null)
        {
            fast=fast.next.next;
            slow=slow.next;
        }


        ListNode prev=null;
        ListNode next=null;
        ListNode curr=slow.next;
        while(curr!=null)
        {
            next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }

        slow.next=null;

        ListNode temp=head;
        fast=temp.next;

        while(temp!=null && prev!=null)
        {
            fast=temp.next;
            curr=prev.next;
            temp.next=prev;
            prev.next=fast;
            prev=curr;
            if(fast==null) break;
            temp=fast;
        }
        if(prev!=null)
        {
            temp.next.next=prev;
        }

    }
}
